src/typechecking.rs

```diff
@@ -0,0 +1,577 @@
+use crate::{proofs::RULE_PROOF_KEYWORD, *};
+
+#[derive(Clone, Debug)]
+pub struct FuncType {
+    pub input: Vec<ArcSort>,
+    pub output: ArcSort,
+}
+
+impl FuncType {
+    pub fn new(input: Vec<ArcSort>, output: ArcSort) -> Self {
+        Self { input, output }
+    }
+}
+
+#[derive(Default, Clone)]
+pub struct TypeInfo {
+    // get the sort from the sorts name()
+    pub presorts: HashMap<Symbol, PreSort>,
+    pub sorts: HashMap<Symbol, Arc<dyn Sort>>,
+    pub primitives: HashMap<Symbol, Vec<Primitive>>,
+    pub func_types: HashMap<Symbol, FuncType>,
+    pub global_types: HashMap<Symbol, ArcSort>,
+    pub local_types: HashMap<CommandId, HashMap<Symbol, ArcSort>>,
```

Comment:
Type info for specific rules- tells what the type of locally-bound variables are.
This new typechecker is redundant with the old one, we should remove the old one eventually